File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
class Solution {
   public:
    int m, n;
    int propagate(int r, int c, vector<vector<int>>& mat) {
        int res = 0;
        int dr = 1;
        while (r + dr < m) {
            if (mat[r + dr][c] == 0) {
                mat[r + dr][c] = 1;
                res++;
            } else if (mat[r + dr][c] > 1)
                break;
            dr++;
        }
        dr = -1;
        while (r + dr >= 0) {
            if (mat[r + dr][c] == 0) {
                mat[r + dr][c] = 1;
                res++;
            } else if (mat[r + dr][c] > 1)
                break;
            dr--;
        }

        int dc = 1;
        while (c + dc < n) {
            if (mat[r][c + dc] == 0) {
                mat[r][c + dc] = 1;
                res++;
            } else if (mat[r][c + dc] > 1)
                break;
            dc++;
        }
        dc = -1;
        while (c + dc >= 0) {
            if (mat[r][c + dc] == 0) {
                mat[r][c + dc] = 1;
                res++;
            } else if (mat[r][c + dc] > 1)
                break;
            dc--;
        }

        return res;
    }
    int countUnguarded(int m, int n, vector<vector<int>>& guards,
                       vector<vector<int>>& walls) {
        this->m = m;
        this->n = n;
        vector<vector<int>> mat(m, vector<int>(n));
        for (const auto& guard : guards) {
            mat[guard[0]][guard[1]] = 2;
        }

        for (const auto& wall : walls) {
            mat[wall[0]][wall[1]] = 3;
        }

        int ans = m * n - guards.size() - walls.size();
        for (const auto& guard : guards) {
            ans -= propagate(guard[0], guard[1], mat);
        }

        return ans;
    }
};

File: 2343-count-unguarded-cells-in-the-grid/2343-count-unguarded-cells-in-the-grid.go
func countUnguarded(m int, n int, guards [][]int, walls [][]int) int {
	mat := make([][]int, m)
	for i := 0; i < m; i++ {
		mat[i] = make([]int, n)
	}

	for _, guard := range guards {
		mat[guard[0]][guard[1]] = 2
	}

	for _, wall := range walls {
		mat[wall[0]][wall[1]] = 3
	}

	ans := m*n - len(guards) - len(walls)
	for _, guard := range guards {
		ans -= propagate(guard[0], guard[1], mat)
	}

	return ans
}

func propagate(r, c int, mat [][]int) int {
	res := 0
	dr := 1
	for r+dr < len(mat) {
		if mat[r+dr][c] == 0 {
			mat[r+dr][c] = 1
			res++
		} else if mat[r+dr][c] > 1 {
			break
		}
		dr++
	}
	dr = -1
	for r+dr >= 0 {
		if mat[r+dr][c] == 0 {
			mat[r+dr][c] = 1
			res++
		} else if mat[r+dr][c] > 1 {
			break
		}
		dr--
	}

	dc := 1
	for c+dc < len(mat[0]) {
		if mat[r][c+dc] == 0 {
			mat[r][c+dc] = 1
			res++
		} else if mat[r][c+dc] > 1 {
			break
		}
		dc++
	}
	dc = -1
	for c+dc >= 0 {
		if mat[r][c+dc] == 0 {
			mat[r][c+dc] = 1
			res++
		} else if mat[r][c+dc] > 1 {
			break
		}
		dc--
	}

	return res
}
